File: sdk/go/spl/spl_test.go
package spl

import (
	"encoding/json"
	"os"
	"strings"
	"testing"
)

// --- Parser tests ---

func TestParseNumber(t *testing.T) {
	n, err := Parse("42")
	if err != nil {
		t.Fatal(err)
	}
	if n != 42.0 {
		t.Fatalf("expected 42, got %v", n)
	}
}

func TestParseNegativeNumber(t *testing.T) {
	n, err := Parse("-3.14")
	if err != nil {
		t.Fatal(err)
	}
	if n != -3.14 {
		t.Fatalf("expected -3.14, got %v", n)
	}
}

func TestParseString(t *testing.T) {
	n, err := Parse(`"hello"`)
	if err != nil {
		t.Fatal(err)
	}
	if n != "hello" {
		t.Fatalf("expected hello, got %v", n)
	}
}

func TestParseBoolTrue(t *testing.T) {
	n, err := Parse("#t")
	if err != nil {
		t.Fatal(err)
	}
	if n != true {
		t.Fatalf("expected true, got %v", n)
	}
}

func TestParseBoolFalse(t *testing.T) {
	n, err := Parse("#f")
	if err != nil {
		t.Fatal(err)
	}
	if n != false {
		t.Fatalf("expected false, got %v", n)
	}
}

func TestParseSymbol(t *testing.T) {
	n, err := Parse("foo")
	if err != nil {
		t.Fatal(err)
	}
	if n != "foo" {
		t.Fatalf("expected foo, got %v", n)
	}
}

func TestParseList(t *testing.T) {
	n, err := Parse(`(and #t #f)`)
	if err != nil {
		t.Fatal(err)
	}
	arr, ok := n.([]Node)
	if !ok {
		t.Fatalf("expected list, got %T", n)
	}
	if len(arr) != 3 {
		t.Fatalf("expected 3 elements, got %d", len(arr))
	}
	if arr[0] != "and" {
		t.Fatalf("expected and, got %v", arr[0])
	}
}

func TestParseNested(t *testing.T) {
	n, err := Parse(`(and (= 1 2) (> 3 1))`)
	if err != nil {
		t.Fatal(err)
	}
	arr := n.([]Node)
	if len(arr) != 3 {
		t.Fatalf("expected 3 elements, got %d", len(arr))
	}
	inner := arr[1].([]Node)
	if inner[0] != "=" {
		t.Fatalf("expected =, got %v", inner[0])
	}
}

func TestParseUnterminatedParen(t *testing.T) {
	_, err := Parse("(and #t")
	if err == nil {
		t.Fatal("expected error for unterminated (")
	}
}

func TestParseUnexpectedCloseParen(t *testing.T) {
	_, err := Parse(")")
	if err == nil {
		t.Fatal("expected error for unexpected )")
	}
}

func TestParseEmptyInput(t *testing.T) {
	_, err := Parse("")
	if err == nil {
		t.Fatal("expected error for empty input")
	}
}

func TestParseStringWithSpaces(t *testing.T) {
	n, err := Parse(`(= "hello world" "hello world")`)
	if err != nil {
		t.Fatal(err)
	}
	arr := n.([]Node)
	if arr[1] != "hello world" {
		t.Fatalf("expected 'hello world', got %v", arr[1])
	}
}

// --- Eval tests ---

func makeEnv() Env {
	env := Env{
		Req: map[string]any{
			"actor_pub":       "K_ai",
			"action":          "payments.create",
			"recipient":       "niece@example.com",
			"purpose":         "giftcard",
			"amount":          50.0,
			"day":             "2025-09-29",
			"device_attested": true,
		},
		Vars: map[string]any{
			"allowed_recipients": []any{"niece@example.com", "mom@example.com"},
			"now":                "2025-10-01T00:00:00Z",
		},
		PerDayCount: func(action, day string) int { return 0 },
	}
	env.Crypto.DPoPOk = func() bool { return true }
	env.Crypto.MerkleOk = func(tuple []any) bool { return true }
	env.Crypto.VRFOk = func(day string, amount float64) bool { return true }
	env.Crypto.ThreshOk = func() bool { return true }
	return env
}

func evalExpr(t *testing.T, src string, env Env) (bool, error) {
	t.Helper()
	ast, err := Parse(src)
	if err != nil {
		t.Fatalf("parse error: %v", err)
	}
	return Verify(ast, env)
}

func TestEvalAndTrue(t *testing.T) {
	ok, err := evalExpr(t, "(and #t #t #t)", makeEnv())
	if err != nil {
		t.Fatal(err)
	}
	if !ok {
		t.Fatal("expected true")
	}
}

func TestEvalAndFalse(t *testing.T) {
	ok, err := evalExpr(t, "(and #t #f #t)", makeEnv())
	if err != nil {
		t.Fatal(err)
	}
	if ok {
		t.Fatal("expected false")
	}
}

func TestEvalOrTrue(t *testing.T) {
	ok, err := evalExpr(t, "(or #f #t #f)", makeEnv())
	if err != nil {
		t.Fatal(err)
	}
	if !ok {
		t.Fatal("expected true")
	}
}

func TestEvalOrFalse(t *testing.T) {
	ok, err := evalExpr(t, "(or #f #f)", makeEnv())
	if err != nil {
		t.Fatal(err)
	}
	if ok {
		t.Fatal("expected false")
	}
}

func TestEvalNot(t *testing.T) {
	ok, err := evalExpr(t, "(not #f)", makeEnv())
	if err != nil {
		t.Fatal(err)
	}
	if !ok {
		t.Fatal("expected true")
	}
}

func TestEvalEqual(t *testing.T) {
	ok, err := evalExpr(t, `(= "hello" "hello")`, makeEnv())
	if err != nil {
		t.Fatal(err)
	}
	if !ok {
		t.Fatal("expected true")
	}
}

func TestEvalEqualFalse(t *testing.T) {
	ok, err := evalExpr(t, `(= "hello" "world")`, makeEnv())
	if err != nil {
		t.Fatal(err)
	}
	if ok {
		t.Fatal("expected false")
	}
}

func TestEvalLessEqual(t *testing.T) {
	ok, err := evalExpr(t, "(<= 5 10)", makeEnv())
	if err != nil {
		t.Fatal(err)
	}
	if !ok {
		t.Fatal("expected true")
	}
}

func TestEvalLessThan(t *testing.T) {
	ok, err := evalExpr(t, "(< 5 5)", makeEnv())
	if err != nil {
		t.Fatal(err)
	}
	if ok {
		t.Fatal("expected false for 5 < 5")
	}
}

func TestEvalGreaterEqual(t *testing.T) {
	ok, err := evalExpr(t, "(>= 10 10)", makeEnv())
	if err != nil {
		t.Fatal(err)
	}
	if !ok {
		t.Fatal("expected true")
	}
}

func TestEvalGreaterThan(t *testing.T) {
	ok, err := evalExpr(t, "(> 10 5)", makeEnv())
	if err != nil {
		t.Fatal(err)
	}
	if !ok {
		t.Fatal("expected true")
	}
}

func TestEvalMember(t *testing.T) {
	env := makeEnv()
	ok, err := evalExpr(t, `(member "niece@example.com" allowed_recipients)`, env)
	if err != nil {
		t.Fatal(err)
	}
	if !ok {
		t.Fatal("expected true")
	}
}

func TestEvalMemberFalse(t *testing.T) {
	env := makeEnv()
	ok, err := evalExpr(t, `(member "stranger@example.com" allowed_recipients)`, env)
	if err != nil {
		t.Fatal(err)
	}
	if ok {
		t.Fatal("expected false")
	}
}

func TestEvalIn(t *testing.T) {
	env := makeEnv()
	ok, err := evalExpr(t, `(in "mom@example.com" allowed_recipients)`, env)
	if err != nil {
		t.Fatal(err)
	}
	if !ok {
		t.Fatal("expected true")
	}
}

func TestEvalSubsetTrue(t *testing.T) {
	env := makeEnv()
	env.Vars["small"] = []any{"a", "b"}
	env.Vars["big"] = []any{"a", "b", "c"}
	ok, err := evalExpr(t, `(subset? small big)`, env)
	if err != nil {
		t.Fatal(err)
	}
	if !ok {
		t.Fatal("expected true")
	}
}

func TestEvalSubsetFalse(t *testing.T) {
	env := makeEnv()
	env.Vars["small"] = []any{"a", "d"}
	env.Vars["big"] = []any{"a", "b", "c"}
	ok, err := evalExpr(t, `(subset? small big)`, env)
	if err != nil {
		t.Fatal(err)
	}
	if ok {
		t.Fatal("expected false")
	}
}

func TestEvalBefore(t *testing.T) {
	env := makeEnv()
	ok, err := evalExpr(t, `(before now "2026-01-01T00:00:00Z")`, env)
	if err != nil {
		t.Fatal(err)
	}
	if !ok {
		t.Fatal("expected true: now is before 2026-01-01")
	}
}

func TestEvalBeforeFalse(t *testing.T) {
	env := makeEnv()
	ok, err := evalExpr(t, `(before now "2025-01-01T00:00:00Z")`, env)
	if err != nil {
		t.Fatal(err)
	}
	if ok {
		t.Fatal("expected false: now is after 2025-01-01")
	}
}

func TestEvalGet(t *testing.T) {
	env := makeEnv()
	ok, err := evalExpr(t, `(= (get req "actor_pub") "K_ai")`, env)
	if err != nil {
		t.Fatal(err)
	}
	if !ok {
		t.Fatal("expected true")
	}
}

func TestEvalPerDayCount(t *testing.T) {
	env := makeEnv()
	ok, err := evalExpr(t, `(<= (per-day-count "payments.create" "2025-09-29") 1)`, env)
	if err != nil {
		t.Fatal(err)
	}
	if !ok {
		t.Fatal("expected true: count 0 <= 1")
	}
}

func TestEvalCryptoStubs(t *testing.T) {
	env := makeEnv()
	for _, expr := range []string{
		"(dpop_ok?)",
		`(merkle_ok? (tuple "a" "b"))`,
		`(vrf_ok? "2025-09-29" 50)`,
		"(thresh_ok?)",
	} {
		ok, err := evalExpr(t, expr, env)
		if err != nil {
			t.Fatalf("expr %q: %v", expr, err)
		}
		if !ok {
			t.Fatalf("expr %q: expected true", expr)
		}
	}
}

func TestEvalUnknownOp(t *testing.T) {
	env := makeEnv()
	_, err := evalExpr(t, `(bogus 1 2)`, env)
	if err == nil {
		t.Fatal("expected error for unknown op")
	}
	if !strings.Contains(err.Error(), "unknown op") {
		t.Fatalf("expected 'unknown op' error, got: %v", err)
	}
}

// --- Gas budget tests ---

func TestGasBudgetExceeded(t *testing.T) {
	env := makeEnv()
	env.MaxGas = 5
	// This policy has more operations than the gas budget
	_, err := evalExpr(t, "(and #t #t #t #t #t #t #t #t #t #t)", env)
	if err == nil {
		t.Fatal("expected gas budget exceeded error")
	}
	if !strings.Contains(err.Error(), "gas budget exceeded") {
		t.Fatalf("expected 'gas budget exceeded', got: %v", err)
	}
}

func TestGasBudgetSufficient(t *testing.T) {
	env := makeEnv()
	env.MaxGas = 100
	ok, err := evalExpr(t, "(and #t #t)", env)
	if err != nil {
		t.Fatal(err)
	}
	if !ok {
		t.Fatal("expected true")
	}
}

// --- Error propagation tests ---

func TestErrorPropagationInAnd(t *testing.T) {
	env := makeEnv()
	_, err := evalExpr(t, `(and #t (bogus))`, env)
	if err == nil {
		t.Fatal("expected error to propagate through and")
	}
}

func TestErrorPropagationInOr(t *testing.T) {
	env := makeEnv()
	_, err := evalExpr(t, `(or #f (bogus))`, env)
	if err == nil {
		t.Fatal("expected error to propagate through or")
	}
}

func TestErrorPropagationInGet(t *testing.T) {
	env := makeEnv()
	// get with non-map first arg should return nil, not panic
	ok, err := evalExpr(t, `(= (get "notamap" "key") "value")`, env)
	if err != nil {
		t.Fatal(err)
	}
	if ok {
		t.Fatal("expected false")
	}
}

// --- Fail-closed crypto defaults ---

func TestCryptoDefaultsFailClosed(t *testing.T) {
	// Env with NO crypto callbacks set — should default to false
	env := Env{
		Req:         map[string]any{},
		Vars:        map[string]any{},
		PerDayCount: func(_, _ string) int { return 0 },
	}
	ok, err := evalExpr(t, "(dpop_ok?)", env)
	if err != nil {
		t.Fatal(err)
	}
	if ok {
		t.Fatal("expected false: crypto defaults should be fail-closed")
	}
}

// --- Strict mode tests ---

func TestStrictModeRejectsUnresolved(t *testing.T) {
	env := makeEnv()
	env.Strict = true
	_, err := evalExpr(t, `(= "foo" unbound_var)`, env)
	if err == nil {
		t.Fatal("expected error for unresolved symbol in strict mode")
	}
	if !strings.Contains(err.Error(), "unresolved symbol") {
		t.Fatalf("expected 'unresolved symbol' error, got: %v", err)
	}
}

func TestNonStrictAllowsUnresolved(t *testing.T) {
	env := makeEnv()
	ok, err := evalExpr(t, `(= "foo" unbound_var)`, env)
	if err != nil {
		t.Fatal(err)
	}
	if ok {
		t.Fatal("expected false: unresolved symbol should not match string")
	}
}

// --- Type-aware equality tests ---

func TestEqualityTypeAware(t *testing.T) {
	env := makeEnv()
	// Number 50 should NOT equal string "50"
	env.Vars["str_fifty"] = "50"
	ok, err := evalExpr(t, `(= 50 str_fifty)`, env)
	if err != nil {
		t.Fatal(err)
	}
	if ok {
		t.Fatal("expected false: number 50 should not equal string '50'")
	}
}

// --- Integration test ---

func TestFamilyGiftsPolicy(t *testing.T) {
	policyBytes, err := os.ReadFile("../../../examples/policies/family_gifts.spl")
	if err != nil {
		t.Skipf("skipping integration test: %v", err)
	}
	reqBytes, err := os.ReadFile("../../../examples/requests/gift_50_niece.json")
	if err != nil {
		t.Skipf("skipping integration test: %v", err)
	}

	ast, err := Parse(string(policyBytes))
	if err != nil {
		t.Fatalf("parse error: %v", err)
	}

	var req map[string]any
	if err := json.Unmarshal(reqBytes, &req); err != nil {
		t.Fatalf("json error: %v", err)
	}

	env := makeEnv()
	env.Req = req

	allow, err := Verify(ast, env)
	if err != nil {
		t.Fatalf("verify error: %v", err)
	}
	if !allow {
		t.Fatal("expected ALLOW for gift_50_niece")
	}
}

func TestFamilyGiftsDeny(t *testing.T) {
	policyBytes, err := os.ReadFile("../../../examples/policies/family_gifts.spl")
	if err != nil {
		t.Skipf("skipping integration test: %v", err)
	}
	reqBytes, err := os.ReadFile("../../../examples/requests/gift_50_niece.json")
	if err != nil {
		t.Skipf("skipping integration test: %v", err)
	}

	ast, err := Parse(string(policyBytes))
	if err != nil {
		t.Fatalf("parse error: %v", err)
	}

	var req map[string]any
	if err := json.Unmarshal(reqBytes, &req); err != nil {
		t.Fatalf("json error: %v", err)
	}

	// Change amount to exceed limit
	req["amount"] = 100.0

	env := makeEnv()
	env.Req = req

	allow, err := Verify(ast, env)
	if err != nil {
		t.Fatalf("verify error: %v", err)
	}
	if allow {
		t.Fatal("expected DENY for amount=100")
	}
}

File: sdk/js/src/spl.test.ts
import { describe, it } from 'node:test';
import * as assert from 'node:assert/strict';
import { parseSExpr, evalPolicy, Sym } from './spl.js';
import { verify } from './verify.js';
import { readFileSync } from 'node:fs';
import { resolve } from 'node:path';

function makeEnv() {
  return {
    vars: {
      allowed_recipients: ['niece@example.com', 'mom@example.com'],
      now: '2025-10-01T00:00:00Z',
    },
    now: '2025-10-01T00:00:00Z',
    per_day_count: (_action: string, _day: string) => 0,
    crypto: {
      dpop_ok: () => true,     // explicit opt-in for tests
      merkle_ok: (_tuple: any) => true,
      vrf_ok: (_day: string, _amount: number) => true,
      thresh_ok: () => true,
    },
  };
}

function evalExpr(src: string, env?: any): boolean {
  const ast = parseSExpr(src);
  const e = env || makeEnv();
  const { allow } = verify(ast, { amount: 0 }, e);
  return allow;
}

// --- Parser tests ---

describe('Parser', () => {
  it('parses numbers', () => {
    assert.equal(parseSExpr('42'), 42);
    assert.equal(parseSExpr('-3.14'), -3.14);
  });

  it('parses strings', () => {
    assert.equal(parseSExpr('"hello"'), 'hello');
  });

  it('parses booleans', () => {
    assert.equal(parseSExpr('#t'), true);
    assert.equal(parseSExpr('#f'), false);
  });

  it('parses symbols', () => {
    const s = parseSExpr('foo');
    assert.ok(s instanceof Sym);
    assert.equal((s as Sym).name, 'foo');
  });

  it('parses lists', () => {
    const ast = parseSExpr('(and #t #f)');
    assert.ok(Array.isArray(ast));
    assert.equal((ast as any[]).length, 3);
    assert.ok((ast as any[])[0] instanceof Sym);
    assert.equal(((ast as any[])[0] as Sym).name, 'and');
  });

  it('parses nested lists', () => {
    const ast = parseSExpr('(and (= 1 2) (> 3 1))');
    assert.ok(Array.isArray(ast));
    const inner = (ast as any[])[1];
    assert.ok(Array.isArray(inner));
    assert.ok(inner[0] instanceof Sym);
    assert.equal((inner[0] as Sym).name, '=');
  });

  it('parses strings with spaces', () => {
    const ast = parseSExpr('(= "hello world" "hello world")');
    assert.equal((ast as any[])[1], 'hello world');
  });

  it('throws on unterminated paren', () => {
    assert.throws(() => parseSExpr('(and #t'), /unterminated/);
  });

  it('throws on unexpected close paren', () => {
    assert.throws(() => parseSExpr(')'), /unexpected/);
  });

  it('throws on extra tokens', () => {
    assert.throws(() => parseSExpr('#t #f'), /extra tokens/);
  });
});

// --- Eval tests ---

describe('Eval', () => {
  it('and - all true', () => {
    assert.equal(evalExpr('(and #t #t #t)'), true);
  });

  it('and - one false', () => {
    assert.equal(evalExpr('(and #t #f #t)'), false);
  });

  it('or - one true', () => {
    assert.equal(evalExpr('(or #f #t #f)'), true);
  });

  it('or - all false', () => {
    assert.equal(evalExpr('(or #f #f)'), false);
  });

  it('not', () => {
    assert.equal(evalExpr('(not #f)'), true);
    assert.equal(evalExpr('(not #t)'), false);
  });

  it('= strings', () => {
    assert.equal(evalExpr('(= "hello" "hello")'), true);
    assert.equal(evalExpr('(= "hello" "world")'), false);
  });

  it('<=', () => {
    assert.equal(evalExpr('(<= 5 10)'), true);
    assert.equal(evalExpr('(<= 10 10)'), true);
    assert.equal(evalExpr('(<= 11 10)'), false);
  });

  it('<', () => {
    assert.equal(evalExpr('(< 5 10)'), true);
    assert.equal(evalExpr('(< 5 5)'), false);
  });

  it('>=', () => {
    assert.equal(evalExpr('(>= 10 10)'), true);
    assert.equal(evalExpr('(>= 9 10)'), false);
  });

  it('>', () => {
    assert.equal(evalExpr('(> 10 5)'), true);
    assert.equal(evalExpr('(> 5 5)'), false);
  });

  it('member', () => {
    assert.equal(evalExpr('(member "niece@example.com" allowed_recipients)'), true);
    assert.equal(evalExpr('(member "stranger@example.com" allowed_recipients)'), false);
  });

  it('in (alias for member)', () => {
    assert.equal(evalExpr('(in "mom@example.com" allowed_recipients)'), true);
  });

  it('subset?', () => {
    const env = makeEnv();
    (env.vars as any).small = ['a', 'b'];
    (env.vars as any).big = ['a', 'b', 'c'];
    assert.equal(evalExpr('(subset? small big)', env), true);
    assert.equal(evalExpr('(subset? big small)', env), false);
  });

  it('before', () => {
    assert.equal(evalExpr('(before now "2026-01-01T00:00:00Z")'), true);
    assert.equal(evalExpr('(before now "2025-01-01T00:00:00Z")'), false);
  });

  it('get', () => {
    const env = makeEnv();
    const ast = parseSExpr('(= (get req "actor_pub") "K_ai")');
    const req = { actor_pub: 'K_ai', action: 'test' };
    const { allow } = verify(ast, req, env);
    assert.equal(allow, true);
  });

  it('crypto callbacks (explicit true)', () => {
    assert.equal(evalExpr('(dpop_ok?)'), true);
    assert.equal(evalExpr('(thresh_ok?)'), true);
  });

  it('crypto defaults are fail-closed', () => {
    const env = {
      vars: {},
      now: '2025-10-01T00:00:00Z',
      per_day_count: (_action: string, _day: string) => 0,
      crypto: {},
    };
    const ast = parseSExpr('(dpop_ok?)');
    const { allow } = verify(ast, {}, env);
    assert.equal(allow, false);
  });

  it('unknown op throws', () => {
    assert.throws(() => evalExpr('(bogus 1 2)'), /Unknown op/);
  });
});

// --- Gas budget tests ---

describe('Gas budget', () => {
  it('exceeds gas budget', () => {
    const env = makeEnv();
    (env as any).maxGas = 3;
    assert.throws(() => evalExpr('(and #t #t #t #t #t #t #t #t)', env), /gas budget exceeded/);
  });

  it('sufficient gas budget', () => {
    const env = makeEnv();
    (env as any).maxGas = 100;
    assert.equal(evalExpr('(and #t #t)', env), true);
  });
});

// --- Strict mode tests ---

describe('Strict mode', () => {
  it('errors on unresolved symbol when strict', () => {
    const env = makeEnv();
    (env as any).strict = true;
    assert.throws(() => evalExpr('(= "foo" unbound_var)', env), /Unresolved symbol: unbound_var/);
  });

  it('allows unresolved symbols when not strict', () => {
    const env = makeEnv();
    // unbound_var resolves to "unbound_var" string, which != "foo"
    assert.equal(evalExpr('(= "foo" unbound_var)', env), false);
  });
});

// --- Integration test ---

describe('Integration', () => {
  it('family_gifts.spl ALLOW', () => {
    let policyBytes: string;
    let reqBytes: string;
    try {
      policyBytes = readFileSync(resolve(import.meta.dirname!, '../../..', 'examples/policies/family_gifts.spl'), 'utf8').trim();
      reqBytes = readFileSync(resolve(import.meta.dirname!, '../../..', 'examples/requests/gift_50_niece.json'), 'utf8');
    } catch {
      return; // skip if files not found
    }

    const ast = parseSExpr(policyBytes);
    const req = JSON.parse(reqBytes);
    const env = makeEnv();
    const { allow } = verify(ast, req, env);
    assert.equal(allow, true);
  });

  it('family_gifts.spl DENY on high amount', () => {
    let policyBytes: string;
    let reqBytes: string;
    try {
      policyBytes = readFileSync(resolve(import.meta.dirname!, '../../..', 'examples/policies/family_gifts.spl'), 'utf8').trim();
      reqBytes = readFileSync(resolve(import.meta.dirname!, '../../..', 'examples/requests/gift_50_niece.json'), 'utf8');
    } catch {
      return; // skip if files not found
    }

    const ast = parseSExpr(policyBytes);
    const req = JSON.parse(reqBytes);
    req.amount = 100;
    const env = makeEnv();
    const { allow } = verify(ast, req, env);
    assert.equal(allow, false);
  });
});
